core/types/state_account_marshalling.go

```diff
@@ -33,73 +32,53 @@ var (
 	ErrInvalidLength = errors.New("StateAccount: invalid input length")
 )
 
-// Hash of StateAccount
-// AccountHash = Hash(
-//	Hash(nonce, balance),
-//  Hash(
-//	  Root,
-//	  Hash(codeHashFirst16, codeHashLast16)
-//  ))
-func (s *StateAccount) Hash() (*big.Int, error) {
-	nonce := new(big.Int).SetUint64(s.Nonce)
-	if s.Balance == nil {
-		s.Balance = new(big.Int)
-	}
-	hash1, err := poseidon.Hash([]*big.Int{nonce, s.Balance})
-	if err != nil {
-		return nil, err
-	}
-
-	codeHashFirst16 := new(big.Int).SetBytes(s.CodeHash[0:16])
-	codeHashLast16 := new(big.Int).SetBytes(s.CodeHash[16:32])
-	hash2, err := poseidon.Hash([]*big.Int{codeHashFirst16, codeHashLast16})
-	if err != nil {
-		return nil, err
-	}
-
-	rootHash := zkt.NewHashFromBytes(s.Root.Bytes())
-	hash3, err := poseidon.Hash([]*big.Int{hash2, rootHash.BigInt()})
-	if err != nil {
-		return nil, err
-	}
-
-	hash4, err := poseidon.Hash([]*big.Int{hash1, hash3})
-	if err != nil {
-		return nil, err
-	}
-	return hash4, nil
-}
-
 // MarshalFields, the bytes scheme is:
 // [0:32] Nonce uint64 big-endian in 32 byte
 // [32:64] Balance
-// [64:96] Root
-// [96:128] CodeHash
+// [64:96] KeccakCodeHash
+// [96:128] Root
+// [128:160] PoseidonCodeHash
+// [160:192] CodeSize
 func (s *StateAccount) MarshalFields() ([]zkt.Byte32, uint32) {
-	fields := make([]zkt.Byte32, 4)
-	binary.BigEndian.PutUint64(fields[0][24:], s.Nonce)
+	fields := make([]zkt.Byte32, 6)
 
 	if !utils.CheckBigIntInField(s.Balance) {
-		panic("balance overflow")
+		panic("StateAccount balance overflow")
 	}
-	s.Balance.FillBytes(fields[1][:])
 
-	copy(fields[2][:], s.CodeHash)
+	binary.BigEndian.PutUint64(fields[0][24:], s.Nonce) // 8-byte value in a 32-byte field
+	s.Balance.FillBytes(fields[1][:])
+	copy(fields[2][:], s.KeccakCodeHash)
 	copy(fields[3][:], s.Root.Bytes())
-	return fields, 4
+	copy(fields[4][:], s.PoseidonCodeHash)
+	binary.BigEndian.PutUint64(fields[5][24:], s.CodeSize) // 8-byte value in a 32-byte field
+
+	// The returned flag shows which items cannot be encoded as a field elements.
+	//
+	// +-------+---------+--------+--------+----------+----------+
+	// | nonce | balance | keccak |  root  | poseidon | codesize |
+	// +-------+---------+--------+--------+----------+----------+
+	//     0        0        1        0         0          0
+
+	flag := uint32(4)
```

Comment:
What is 4 with regard to "which items cannot be encoded as a field elements"?

Reply:
It's a bitfield. `4 = 001000` (LSB first) means that `keccak` uses a special encoding (since its values often don't fit into a single field element).

Reply:
<img width="1224" alt="image" src="https://user-images.githubusercontent.com/7571518/215709901-d64d10d5-8569-4db9-bff2-5db10a852462.png">